tsdb/aggregate.go

```diff
@@ -0,0 +1,892 @@
+package tsdb
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/influxdb/influxdb/influxql"
+	"github.com/influxdb/influxdb/models"
+	"github.com/influxdb/influxdb/pkg/slices"
+)
+
+// AggregateExecutor represents a mapper for execute aggregate SELECT statements.
+type AggregateExecutor struct {
+	stmt    *influxql.SelectStatement
+	mappers []*StatefulMapper
+}
+
+// NewAggregateExecutor returns a new AggregateExecutor.
+func NewAggregateExecutor(stmt *influxql.SelectStatement, mappers []Mapper) *AggregateExecutor {
+	e := &AggregateExecutor{
+		stmt:    stmt,
+		mappers: make([]*StatefulMapper, 0, len(mappers)),
+	}
+
+	for _, m := range mappers {
+		e.mappers = append(e.mappers, &StatefulMapper{m, nil, false})
+	}
+
+	return e
+}
+
+// close closes the executor such that all resources are released.
+// Once closed, an executor may not be re-used.
+func (e *AggregateExecutor) close() {
+	if e != nil {
+		for _, m := range e.mappers {
+			m.Close()
+		}
+	}
+}
+
+// Execute begins execution of the query and returns a channel to receive rows.
+func (e *AggregateExecutor) Execute() <-chan *models.Row {
+	out := make(chan *models.Row, 0)
+	go e.execute(out)
+	return out
+}
+
+func (e *AggregateExecutor) execute(out chan *models.Row) {
+	// It's important to close all resources when execution completes.
+	defer e.close()
+
+	// Create the functions which will reduce values from mappers for
+	// a given interval. The function offsets within this slice match
+	// the offsets within the value slices that are returned by the
+	// mapper.
+	reduceFuncs, err := e.initReduceFuncs()
+	if err != nil {
+		out <- &models.Row{Err: err}
+		return
+	}
+
+	// Put together the rows to return, starting with columns.
+	columnNames := e.stmt.ColumnNames()
+
+	// Open the mappers.
+	if err := e.openMappers(); err != nil {
+		out <- &models.Row{Err: err}
+		return
+	}
+
+	// Filter out empty sets if there are multiple tag sets.
+	hasMultipleTagSets := e.hasMultipleTagSets()
+	ascending := e.ascending()
+
+	// Prime each mapper's chunk buffer.
+	if err := e.initMappers(); err != nil {
+		out <- &models.Row{Err: err}
+		return
+	}
+
+	// Keep looping until all mappers drained.
+	for !e.mappersDrained() {
+		chunks, err := e.readNextTagset()
+		if err != nil {
+			out <- &models.Row{Err: err}
+			return
+		}
+
+		// Prep a row, ready for kicking out.
+		row := &models.Row{
+			Name:    chunks[0].Name,
+			Tags:    chunks[0].Tags,
+			Columns: columnNames,
+		}
+
+		// Prep for bucketing data by start time of the interval.
+		buckets := map[int64][][]interface{}{}
+
+		var chunkValues []*MapperValue
+		for _, chunk := range chunks {
+			for _, chunkValue := range chunk.Values {
```

Comment:
Yeah, the code isn't optimized. It's primarily a refactoring to remove the `TagSetCursor` code and still make all the tests pass. I think there's a bunch of places for optimization.